### ISSUE TITLE
PorousFlow: initial condition for persistent variable model

### ISSUE THREAD

**Issue by cpgr** [idaholab/moose#9377]
### Description of the enhancement or error report
Add an initial condition to allow a phase saturation to be specified for the persistent variable models.

### Rationale for the enhancement or information for reproducing the error
In many cases, a user will want to specify an initial saturation of one of the phases. However, the complex fluid state models in porous flow that calculate the partitioning of each fluid component in each phase use pressure and total mass fraction of a fluid component summed over all phases as the primary nonlinear variables.

In order to specify an initial saturation, the value of total mass fraction that gives that saturation must be calculated. This calculation can be done in an initial condition, where saturation is specified and the value of total mass fraction is calculated and set.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
New capability that will make it easier to do problems where the initial saturation is known.

**Comment by cpgr** [comment]
This is something that I have started and seems simple enough. However, I need to use a heap of code in one of our materials, and I'm not sure of the best way to get at it.

For example, I need access to the `massFraction()` method that is currently a protected method in the `PorousFlowFluidStateBrineCO2` material (which needs 300 lines of code from here https://github.com/idaholab/moose/blob/devel/modules/porous_flow/src/materials/PorousFlowFluidStateBrineCO2.C#L339).

I don't think that I can access it as is, so these methods will need to be moved to a namespace or a class somewhere as I certainly don't want to duplicate this much code! Any preferences (namespace or class) or other ideas?

**Comment by andrsd** [comment]
You want to go with a class that you will specify in your input file. This object will provide an API to the client side. Then you can inherit from that and implement your BrineCO2 mixture. Similar design to what we do in `fluid_properties` module.  Then your material will take your new user object and it will be named something like `PorousFlowFluidStateMaterial` (similar to our `FluidPropertiesMaterial` in `fluid_properties` module) and you will have just one of those for all possible combinations.

**Comment by cpgr** [comment]
I've had a go at moving all of the actual logic from these materials to user objects, but I'm not sure that I can actually replace the `PorousFlowFluidStateFoo` materials with a single one because they take a different number of coupled variables. `PorousFlowFluidStateWaterNCG` expects `p` and `z`, while `PorousFlowFluidStateBrineCO2` needs `p`, `z`, and `xnacl`.

So `PorousFlowFluidStateWaterNCG` would call `someFunction(p,z)` but `PorousFlowFluidStateBrineCO2` needs to call `someFunction(p,z,xnacl)`

I have uploaded what I have done so far to https://github.com/cpgr/moose/commit/32b965ba98c13719c4d93e5c487d214e339e6894 

On the plus side, even if I can't reduce the number of materials, this change has allowed me to write some unit tests that I couldn't before, which meant that I fixed a pretty subtle error in one of the derivatives.

**Comment by andrsd** [comment]
The functions in your user object can take an additional parameter - mass fraction. And then the classes that do not have it would supply some fake value.  This may or may not be the right solution, though.  We had a similar design decision in RELAP-7, where HEM was implemented as single phase + one extra parameter, so that fubar'd all single phase kernels, materials, etc. In the end we separated the two even though they had quite a bit in common. But in the end, it allowed for better error checking and simpler classes. Not sure if this would apply to this particular case.  But, these are your 2 basic options.
